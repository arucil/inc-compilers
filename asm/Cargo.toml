[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
