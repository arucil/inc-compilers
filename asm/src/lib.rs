#![feature(never_type)]

use num_derive::{FromPrimitive, ToPrimitive};
use std::fmt::{self, Debug, Formatter, Write};

#[derive(Debug, Clone)]
pub struct Program<INFO = (), VAR = !> {
  pub info: INFO,
  pub blocks: Vec<(String, Block<VAR>)>,
}

#[derive(Clone)]
pub struct Block<VAR = !> {
  pub code: Vec<Instr<VAR>>,
}

#[derive(Clone)]
#[non_exhaustive]
pub enum Instr<VAR = !> {
  Add(Arg<VAR>, Arg<VAR>),
  Sub(Arg<VAR>, Arg<VAR>),
  Mov(Arg<VAR>, Arg<VAR>),
  Neg(Arg<VAR>),
  Call(String, usize),
  Ret,
  Push(Arg<VAR>),
  Pop(Arg<VAR>),
  Jmp(String),
  Syscall,
}

#[derive(Clone, PartialEq, Eq)]
pub enum Arg<VAR = !> {
  Imm(i64),
  Reg(Reg),
  Deref(Reg, i32),
  Var(VAR),
}

#[derive(
  Clone, Copy, FromPrimitive, ToPrimitive, PartialEq, Eq, PartialOrd, Ord, Hash,
)]
pub enum Reg {
  Rsp,
  Rbp,
  Rax,
  Rbx,
  Rcx,
  Rdx,
  Rsi,
  Rdi,
  R8,
  R9,
  R10,
  R11,
  R12,
  R13,
  R14,
  R15,
}

impl<INFO: Debug, VAR: Debug> Program<INFO, VAR> {
  pub fn to_string_pretty(&self) -> String {
    let mut buf = format!("{:?}", self.info);
    for (label, block) in &self.blocks {
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }

  pub fn to_nasm(&self) -> String {
    let mut buf = format!(
      "extern read_int, print_int, print_newline\n\
      section .text\n"
    );
    for (label, block) in &self.blocks {
      buf += "\n";
      if label == "_start" {
        buf += "    global _start\n";
      }
      writeln!(&mut buf, "{}:", label).unwrap();
      buf += &format!("{:?}", block);
    }
    buf
  }
}

impl<VAR: Debug> Debug for Block<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    for instr in &self.code {
      write!(f, "    ")?;
      instr.fmt(f)?;
      writeln!(f)?;
    }
    Ok(())
  }
}

impl<VAR: Debug> Debug for Instr<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Add(src, dest) => write!(f, "add {:?}, {:?}", dest, src),
      Self::Mov(src, dest) => write!(f, "mov {:?}, {:?}", dest, src),
      Self::Call(label, _) => write!(f, "call {}", label),
      Self::Jmp(label) => write!(f, "jmp {}", label),
      Self::Neg(dest) => write!(f, "neg {:?}", dest),
      Self::Pop(dest) => write!(f, "pop {:?}", dest),
      Self::Push(src) => write!(f, "push {:?}", src),
      Self::Ret => write!(f, "ret"),
      Self::Sub(src, dest) => write!(f, "sub {:?}, {:?}", dest, src),
      Self::Syscall => write!(f, "syscall"),
    }
  }
}

impl<VAR: Debug> Debug for Arg<VAR> {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    match self {
      Self::Imm(n) => write!(f, "{}", n),
      Self::Var(var) => var.fmt(f),
      Self::Deref(r, i) => {
        if *i > 0 {
          write!(f, "qword [{:?} + {}]", r, i)
        } else if *i == 0 {
          write!(f, "qword [{:?}]", r)
        } else {
          write!(f, "qword [{:?} - {}]", r, -i)
        }
      }
      Self::Reg(r) => r.fmt(f),
    }
  }
}

impl Reg {
  pub fn is_callee_saved(&self) -> bool {
    match self {
      Reg::Rsp
      | Reg::Rbp
      | Reg::Rbx
      | Reg::R12
      | Reg::R13
      | Reg::R14
      | Reg::R15 => true,
      _ => false,
    }
  }
}

impl Debug for Reg {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    use Reg::*;
    let op = match self {
      Rsp => "rsp",
      Rbp => "rbp",
      Rax => "rax",
      Rbx => "rbx",
      Rcx => "rcx",
      Rdx => "rdx",
      Rsi => "rsi",
      Rdi => "rdi",
      R8 => "r8",
      R9 => "r9",
      R10 => "r10",
      R11 => "r11",
      R12 => "r12",
      R13 => "r13",
      R14 => "r14",
      R15 => "r15",
    };
    f.write_str(op)
  }
}
