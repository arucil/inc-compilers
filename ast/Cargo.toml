[package]
name = "ast"
version = "0.1.0"
edition = "2018"

[dependencies]
pretty = "0.10.0"
support = { path = "../support" }

[dev-dependencies]
pretty_assertions = "0.7.2"
