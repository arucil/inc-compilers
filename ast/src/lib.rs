#![feature(never_type, box_syntax)]

use id_arena::{Arena, Id};
use pretty::*;
use std::fmt::{self, Debug, Formatter};
use std::iter;
use support::Range;

pub mod parser;

pub use parser::{parse, Result};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program<VAR = String, TYPE = ()> {
  pub body: Vec<Exp<VAR, TYPE>>,
  pub types: Arena<CompType>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Exp<VAR = String, TYPE = ()> {
  pub kind: ExpKind<VAR, TYPE>,
  pub range: Range,
  pub ty: TYPE,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ExpKind<VAR, TYPE> {
  Int(i64),
  Prim {
    op: (Range, &'static str),
    args: Vec<Exp<VAR, TYPE>>,
  },
  Var(VAR),
  Str(String),
  Let {
    var: (Range, VAR),
    init: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Bool(bool),
  If {
    cond: Box<Exp<VAR, TYPE>>,
    conseq: Box<Exp<VAR, TYPE>>,
    alt: Box<Exp<VAR, TYPE>>,
  },
  Set {
    var: (Range, VAR),
    exp: Box<Exp<VAR, TYPE>>,
  },
  Get(VAR),
  Begin {
    seq: Vec<Exp<VAR, TYPE>>,
    last: Box<Exp<VAR, TYPE>>,
  },
  While {
    cond: Box<Exp<VAR, TYPE>>,
    body: Box<Exp<VAR, TYPE>>,
  },
  Void,
  Print(
    /// nonempty
    Vec<Exp<VAR, TYPE>>,
  ),
  NewLine,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Type {
  Int,
  Bool,
  Str,
  Comp(TypeId),
  Void,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TypeId(Id<CompType>);

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompType {
  Vector(Vec<Type>),
}

#[derive(Clone, PartialEq, Eq, Hash)]
pub struct IdxVar {
  pub name: String,
  pub index: usize,
}

impl IdxVar {
  pub fn new<S: ToString>(name: S) -> Self {
    let name = name.to_string();
    if let Some(i) = name.find('.') {
      if let Ok(index) = name[i + 1..].parse::<usize>() {
        return Self {
          name: name[..i].to_owned(),
          index,
        };
      }
    }
    Self { name, index: 0 }
  }
}

impl Debug for IdxVar {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    write!(f, "{}.{}", self.name, self.index)
  }
}

impl<VAR: Debug, TYPE: Debug> Program<VAR, TYPE> {
  pub fn to_string_pretty(&self) -> String {
    let doc = self.to_doc();
    let mut buf = String::new();
    doc.render_fmt(30, &mut buf).unwrap();
    buf
  }

  fn to_doc(&self) -> RcDoc {
    RcDoc::intersperse(self.body.iter().map(|exp| exp.to_doc()), Doc::line())
  }
}

impl<VAR: Debug, TYPE: Debug> Exp<VAR, TYPE> {
  fn to_doc(&self) -> RcDoc {
    match &self.kind {
      ExpKind::Int(n) => RcDoc::text(format!("{}", n)),
      ExpKind::Str(s) => RcDoc::text(format!("{:?}", s)),
      ExpKind::Var(var) | ExpKind::Get(var) => {
        RcDoc::text(format!("{:?}", var))
      }
      ExpKind::Prim { op, args } => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text(op.1))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Let { var, init, body } => RcDoc::text("(")
        .append(
          RcDoc::text("let")
            .append(Doc::line())
            .append(
              RcDoc::text("[")
                .append(
                  RcDoc::text(format!("{:?}", var.1))
                    .append(Doc::line())
                    .append(init.to_doc())
                    .nest(1)
                    .group(),
                )
                .append(RcDoc::text("]")),
            )
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Bool(true) => RcDoc::text("#t"),
      ExpKind::Bool(false) => RcDoc::text("#f"),
      ExpKind::If { cond, conseq, alt } => RcDoc::text("(")
        .append(
          RcDoc::text("if")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(conseq.to_doc().append(Doc::line()).append(alt.to_doc()))
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Set { var, exp } => RcDoc::text("(")
        .append(
          RcDoc::text("set!")
            .append(Doc::line())
            .append(RcDoc::text(format!("{:?}", var.1)))
            .append(Doc::line())
            .append(exp.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Begin { seq, last } => RcDoc::text("(")
        .append(
          RcDoc::text("begin").append(Doc::line()).append(
            RcDoc::intersperse(
              seq
                .iter()
                .chain(std::iter::once(&**last))
                .map(|exp| exp.to_doc()),
              Doc::line(),
            )
            .nest(1)
            .group(),
          ),
        )
        .append(RcDoc::text(")")),
      ExpKind::While { cond, body } => RcDoc::text("(")
        .append(
          RcDoc::text("while")
            .append(Doc::line())
            .append(cond.to_doc())
            .group()
            .append(Doc::line())
            .append(body.to_doc())
            .nest(1)
            .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::Void => RcDoc::text("(void)"),
      ExpKind::Print(args) => RcDoc::text("(")
        .append(
          RcDoc::intersperse(
            iter::once(RcDoc::text("print"))
              .chain(args.iter().map(|arg| arg.to_doc())),
            Doc::line(),
          )
          .nest(1)
          .group(),
        )
        .append(RcDoc::text(")")),
      ExpKind::NewLine => RcDoc::text("(newline)"),
    }
  }
}
