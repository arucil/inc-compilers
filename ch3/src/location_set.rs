use asm::{Arg, Reg};
use ast::IdxVar;
use indexmap::map::Entry;
use indexmap::IndexMap;
use num_traits::{FromPrimitive, ToPrimitive};
use smallvec::SmallVec;
use std::fmt::{self, Debug, Write};
use std::ops::Deref;

#[derive(Debug, Clone)]
pub struct LocationSet(SmallVec<[u32; 2]>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Location(usize);

#[derive(Debug, Clone)]
pub struct VarStore(IndexMap<IdxVar, Var>);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Var(usize);

impl LocationSet {
  pub fn new() -> Self {
    Self(SmallVec::new())
  }

  pub fn add_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if self.0.is_empty() {
      self.0.push(0);
    }
    self.0[i / 32] |= 1 << (i % 32);
  }

  pub fn remove_reg(&mut self, reg: Reg) {
    let i = reg_index(reg);
    if !self.0.is_empty() {
      self.0[i / 32] &= !(1 << (i % 32));
    }
  }

  pub fn add_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset >= self.0.len() {
      self.0.resize(offset + 1, 0);
    }
    self.0[offset] |= 1 << (i % 32);
  }

  pub fn remove_var(&mut self, var: Var) {
    let i = var_index(var);
    let offset = i / 32;
    if offset < self.0.len() {
      self.0[offset] &= !(1 << (i % 32));
    }
  }

  pub fn remove_caller_saved_regs(&mut self) {
    self.remove_reg(Reg::Rax);
    self.remove_reg(Reg::Rcx);
    self.remove_reg(Reg::Rdx);
    self.remove_reg(Reg::Rsi);
    self.remove_reg(Reg::Rdi);
    self.remove_reg(Reg::R8);
    self.remove_reg(Reg::R9);
    self.remove_reg(Reg::R10);
    self.remove_reg(Reg::R11);
  }

  pub fn add_argument_regs(&mut self, arity: usize) {
    const ARGUMENT_REGS: [Reg; 6] =
      [Reg::Rdi, Reg::Rsi, Reg::Rdx, Reg::Rcx, Reg::R8, Reg::R9];

    assert!(arity <= 6);
    for reg in &ARGUMENT_REGS[..arity] {
      self.add_reg(*reg);
    }
  }

  pub fn iter(&self) -> LocationIter {
    let current = if self.0.len() > 0 { self.0[0] } else { 0 };
    LocationIter {
      set: self,
      index: 0,
      current,
    }
  }

  pub fn write(&self, f: &mut impl Write, var_store: &VarStore) -> fmt::Result {
    write!(f, "{{")?;
    let mut comma = false;
    for elem in self {
      if comma {
        write!(f, ", ")?;
      }
      comma = true;
      write!(f, "{:?}", elem.to_arg(var_store))?;
    }
    write!(f, "}}")
  }
}

pub struct LocationIter<'a> {
  set: &'a LocationSet,
  index: usize,
  current: u32,
}

impl<'a> Iterator for LocationIter<'a> {
  type Item = Location;

  fn next(&mut self) -> Option<Location> {
    while self.current == 0 {
      self.index += 1;
      if self.index >= self.set.0.len() {
        return None;
      }
      self.current = self.set.0[self.index];
    }
    let i = self.current.trailing_zeros() as usize + self.index * 32;
    self.current &= self.current - 1;
    Some(Location(i))
  }
}

impl<'a> IntoIterator for &'a LocationSet {
  type Item = Location;
  type IntoIter = LocationIter<'a>;

  fn into_iter(self) -> Self::IntoIter {
    self.iter()
  }
}

fn reg_index(reg: Reg) -> usize {
  reg.to_usize().unwrap()
}

fn var_index(var: Var) -> usize {
  var.0 + 16
}

impl VarStore {
  pub fn new() -> Self {
    Self(IndexMap::new())
  }

  pub fn get(&mut self, var: IdxVar) -> Var {
    let len = self.0.len();
    match self.0.entry(var) {
      Entry::Occupied(var) => *var.get(),
      Entry::Vacant(entry) => *entry.insert(Var(len)),
    }
  }
}

impl Deref for VarStore {
  type Target = IndexMap<IdxVar, Var>;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

impl Var {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    Arg::Var(var_store.get_index(self.0 - 16).unwrap().0.clone())
  }
}

impl Location {
  pub fn to_arg(&self, var_store: &VarStore) -> Arg<IdxVar> {
    if self.0 < 16 {
      Arg::Reg(Reg::from_usize(self.0).unwrap())
    } else {
      Arg::Var(var_store.0.get_index(self.0 - 16).unwrap().0.clone())
    }
  }

  pub fn from_arg(arg: Arg<IdxVar>, var_store: &mut VarStore) -> Option<Self> {
    match arg {
      Arg::Deref(reg, _) | Arg::Reg(reg) => Some(reg.into()),
      Arg::Var(var) => Some(Self(var_store.get(var).0 + 16)),
      Arg::Imm(_) => None,
      Arg::ByteReg(_) => unimplemented!(),
    }
  }

  pub fn to_reg(&self) -> Option<Reg> {
    Reg::from_usize(self.0)
  }

  pub fn to_var(&self) -> Option<Var> {
    if self.0 >= 16 {
      Some(Var(self.0 - 16))
    } else {
      None
    }
  }
}

impl From<Reg> for Location {
  fn from(reg: Reg) -> Self {
    Self(reg.to_usize().unwrap())
  }
}
