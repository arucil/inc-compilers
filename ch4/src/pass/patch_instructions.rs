use asm::{Arg, Block, Instr, Program, Reg};

pub fn patch_instructions<T>(prog: Program<T>) -> Program<T> {
  Program {
    info: prog.info,
    constants: prog.constants,
    blocks: prog
      .blocks
      .into_iter()
      .map(|(label, block)| (label, patch_block(block)))
      .collect(),
  }
}

fn patch_block(block: Block) -> Block {
  let mut code = vec![];
  for instr in block.code {
    match instr {
      Instr::Add {
        src: src @ Arg::Deref(..),
        dest: dest @ Arg::Deref(..),
      } => {
        code.push(Instr::Mov {
          src,
          dest: Arg::Reg(Reg::Rax),
        });
        code.push(Instr::Add {
          src: Arg::Reg(Reg::Rax),
          dest,
        });
      }
      Instr::Sub {
        src: src @ Arg::Deref(..),
        dest: dest @ Arg::Deref(..),
      } => {
        code.push(Instr::Mov {
          src,
          dest: Arg::Reg(Reg::Rax),
        });
        code.push(Instr::Sub {
          src: Arg::Reg(Reg::Rax),
          dest,
        });
      }
      Instr::Xor {
        src: src @ Arg::Deref(..),
        dest: dest @ Arg::Deref(..),
      } => {
        code.push(Instr::Mov {
          src,
          dest: Arg::Reg(Reg::Rax),
        });
        code.push(Instr::Xor {
          src: Arg::Reg(Reg::Rax),
          dest,
        });
      }
      Instr::Cmp {
        src: src @ Arg::Deref(..),
        dest: dest @ Arg::Deref(..),
      } => {
        code.push(Instr::Mov {
          src,
          dest: Arg::Reg(Reg::Rax),
        });
        code.push(Instr::Cmp {
          src: Arg::Reg(Reg::Rax),
          dest,
        });
      }
      Instr::Cmp {
        src,
        dest: dest @ Arg::Imm(_),
      } => {
        code.push(Instr::Mov {
          src: dest,
          dest: Arg::Reg(Reg::Rax),
        });
        code.push(Instr::Cmp {
          src,
          dest: Arg::Reg(Reg::Rax),
        });
      }
      Instr::Mov {
        src: src @ Arg::Deref(..),
        dest: dest @ Arg::Deref(..),
      } => {
        if src != dest {
          code.push(Instr::Mov {
            src,
            dest: Arg::Reg(Reg::Rax),
          });
          code.push(Instr::Mov {
            src: Arg::Reg(Reg::Rax),
            dest,
          });
        }
      }
      Instr::Mov {
        src: Arg::Reg(ref src),
        dest: Arg::Reg(ref dest),
      } => {
        if src != dest {
          code.push(instr);
        }
      }
      Instr::Mov {
        src: Arg::Imm(0),
        dest: dest @ Arg::Reg(_),
      } => {
        code.push(Instr::Xor {
          src: dest.clone(),
          dest,
        });
      }
      Instr::Movzx { src, dest } if !matches!(dest, Arg::Reg(_)) => {
        code.push(Instr::Movzx {
          src,
          dest: Arg::Reg(Reg::Rax),
        });
        code.push(Instr::Mov {
          src: Arg::Reg(Reg::Rax),
          dest,
        })
      }
      _ => {
        code.push(instr);
      }
    }
  }
  Block {
    global: block.global,
    code,
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use asm::{Label, Reg::*};
  use ast::*;
  use ch3::location_set::LocationSet;
  use insta::assert_snapshot;
  use maplit::hashmap;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = anf::anf(prog);
    let prog = explicate_control::explicate_control(prog);
    let prog = select_instruction::select_instruction(prog);
    let prog = liveness_analysis::analyze_liveness(
      prog,
      hashmap! {
        Label::Conclusion => LocationSet::regs([Rax, Rbp])
      },
    );
    let prog = interference::build_interference(prog);
    let prog = ch3::pass::move_biasing::build_move_graph(prog);
    let regs = &[
      Rbx, Rcx, Rdx, Rsi, Rdi, R8, R9, R10, R11, R12, R13, R14, R15,
    ];
    let prog = ch3::pass::register_allocation::allocate_registers(prog, regs);
    let result = patch_instructions(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
