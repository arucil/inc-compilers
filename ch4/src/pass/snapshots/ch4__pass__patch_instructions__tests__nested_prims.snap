---
source: ch4/src/pass/patch_instructions.rs
expression: result.to_string_pretty()
---
locals: {x.0, y.1, tmp.0, tmp.1, tmp.2, tmp.3}
used_callee_saved_regs: {rbx}
stack_space: 0 bytes
constants: {}

start:
    call read_int
    mov rbx, rax
    mov rbx, 2
    add rbx, 3
    call read_int
    mov rcx, rax
    neg rcx
    mov rdx, 2
    neg rdx
    add rbx, rdx
    mov rax, rcx
    add rax, rbx
    jmp conclusion

